### ISSUE TITLE
[Question] Can jotai be used without Suspense?

### ISSUE THREAD

**Issue by dai-shi** [pmndrs/jotai#269]
In @JLarky 's [article](https://blog.logrocket.com/jotai-vs-recoil-what-are-the-differences/), it notes:

> The first set of features that stands out are functions that help to work with async selectors. For example, if you want to work with an async atom in a context where Suspense would not be appropriate, you can always consume it with useRecoilStateLoadable or useRecoilValueLoadable so it will no longer throw errors or promises.

The question is if we can use jotai without Suspense. Actually, we can under some constraints.

There are two parts it requires Suspense.

- Derived async atoms (async `read`)
<img width="481" alt="image" src="https://user-images.githubusercontent.com/490574/105330884-24b7e700-5c16-11eb-9e54-e299727edcb8.png">

- Async actions (async `write`)
<img width="406" alt="image" src="https://user-images.githubusercontent.com/490574/105330845-1a95e880-5c16-11eb-82d9-916acbc77366.png">

---

The latter is easier.

The original code is this:
```js
const fetchCountAtom = atom(
  get => get(countAtom),
  async (_get, set, url) => {
    const response = await fetch(url)
    set(countAtom, (await response.json()).count)
  }
)
```
The `write` function (2nd arg) returns a promise, which will suspend.

So, We just don't return anything.
The modified code is this:
```js
const fetchCountAtom = atom(
  get => get(countAtom),
  (_get, set, url) => {
    const fetchData = async () => {
      const response = await fetch(url)
      set(countAtom, (await response.json()).count)
    }
    fetchData()
  }
)
```

---

Okay, the difficult one is async read.
The original code is this:
```js
const urlAtom = atom("https://json.host.com")
const fetchUrlAtom = atom(
  async get => {
    const response = await fetch(get(urlAtom))
    return await response.json()
  }
)
```

We can't make `read` function to return a promise.
So, we have to change the structure.
Here's the new code:
```js
const fetchResultAtom = atom({ loading: true, error: null, data: null })
const runFetchAtom = atom(null, (_get, set, url) => {
  const fetchData = async () => {
    set(fetchResultAtom, prev => ({ ...prev, loading: true }))
    try {
      const response = await fetch(url)
      const data = await response.json()
      set(fetchResultAtom, { loading: false, error: null, data })
    } catch (error) {
      set(fetchResultAtom, { loading: false, error, data: null })
    }
  }
  fetchData()
})

const Component = () => {
  const [, runFetch] = useAtom(runFetchAtom)
  useEffect(() => {
    runFetch("https://json.host.com")
  }, [])
```
Yes, we need useEffect for initialization. This would be mitigated once atom effects #211 lands.

**Comment by allforabit** [comment]
Here's what I came up with to run a promise without suspense based on your example:
```typescript
export const atomWithPromise = <T>(promiseFn: (get: Getter) => Promise<T>) => {
  const promResultAtom = atom<any>({ loading: true, error: null, data: null });

  const runFetchAtom = atom(
    (get) => get(promResultAtom),
    (get, set) => {
      const fetchData = async () => {
        set(promResultAtom, (prev: any) => ({ ...prev, loading: true }));
        try {
          const data = await promiseFn(get);
          set(promResultAtom, { loading: false, error: null, data });
        } catch (error) {
          console.log(error, "error");
          set(promResultAtom, { loading: false, error, data: null });
        }
      };
      fetchData();
    }
  );

  runFetchAtom.onMount = (send) => {
    // Trigger fetch atom
    console.log("Mount");
    send();
  };

  const readonlyAtom = atom((get) => get(runFetchAtom));
  return readonlyAtom;
};
```

**Comment by dai-shi** [comment]
@allforabit Nice!!!
Curious if you don't need to support refetching?

**Comment by allforabit** [comment]
Thanks @dai-shi 

For now I don't need it but will post an update if I add it in. I only refetch when the dependency changes (which creates a new promise atom).

**Comment by dai-shi** [comment]
> Yes, we need useEffect for initialization. This would be mitigated once atom effects #211 lands.

We have `onMount` now, so the best practice is slightly changed.

**Comment by toburger** [comment]
I need to refetch the data every time an atom dependency changes. The above implementation doesn't work for such a scenario as writes are not tracked.
Any idea on how to get it to work?

Another difference to `useRecoilStateLoadable` is the state machine approach. There exist a plethora of components that are using a loading prop to signal a "busy" component. Would be nice to have simple support for those components with jotai.

**Comment by dai-shi** [comment]
Yeah, I wrote this https://docs.pmnd.rs/jotai/guides/no-suspense#async-read
but found it uncomfortable.

Now, as you chime in, I re-consider this...

And, came up with this workaround!!!
https://codesandbox.io/s/blissful-surf-zulsl?file=/src/App.js

---

> Would be nice to have simple support for those components with jotai.

We believe Suspense is the future we should aim at, and the jotai core is fully designed for it.
However, a util to support this workaround would be possible. `lodableAtom` or something. In this case, we wouldn't want to throw error either.

**Comment by dai-shi** [comment]
> Warning: Maximum update depth exceeded. This can happen when a component calls setState inside useEffect, but useEffect either doesn't have a dependency array, or one of the dependencies changes on every render.

Oops...

**Comment by Pinpickle** [comment]
@dai-shi I ran into this issue myself. I really want to have some kind of `useRecoilValueLoadable`, and still be able to use suspense where it makes sense. Suspense isn't ideal 100% of the time as it kills the component tree.

Looking at your example [here](https://github.com/pmndrs/jotai/issues/269#issuecomment-873563552), I ran into the same maximum update depth warning. It turns out if the atom returns something referentially equal, the warning goes away:

So change this

```
const dataAtom = atom((get) => {
  try {
    return { data: get(asyncCountAtom) };
  } catch (e) {
    if (e instanceof Promise) {
      return { loading: true };
    }
    throw e;
  }
});
```

To this

```
const LOADING = { loading: true };

const dataAtom = atom((get) => {
  try {
    return { data: get(asyncCountAtom) };
  } catch (e) {
    if (e instanceof Promise) {
      return LOADING;
    }
    throw e;
  }
});
```

I'm not really sure what Jotai internals cause this phenomenon, but I hope this helps.

**Comment by dai-shi** [comment]
@Pinpickle 
Wow, you resolved it!
https://codesandbox.io/s/stoic-feistel-qo849?file=/src/App.js
Yes, this looks like a good solution.

I think this is a good workaround.
We can add it to https://docs.pmnd.rs/jotai/guides/no-suspense.

We could even consider adding a new util.
```js
import { loadable } from 'jotai/utils'

const anAtom = atom(aync (get) => { ... })
const anAtomThatDoesntThrowPromise = loadable(anAtom) // { data, loading, error }
```
Maybe it should handle error too, instead of throwing to ErrorBoundary.
The implementation would be somewhat close to `waitFor`. (maybe not so close, probably simpler, and this time it's writable.)

How does that sound? Would you be interested in working on it?

**Comment by Pinpickle** [comment]
@dai-shi I'd love to have a go working on it! I think I might be able to sometime in the next couple of weeks.

We could make `loadable` cache the result in a `WeakMap`, so this becomes possible:

```
const value = useAtomValue(loadable(asyncAtom));
```

What do you think about also having a `useLoadableAtomValue` hook that automatically creates a 

Personally I'm not sure about the error boundary use-case, I can image it being useful to throw or not throw depending on context. What about having two different functions?

```
const SUSPENDED = new Symbol("jotai/atomSuspended");

type Loadable<T, TError = any> = { type: "loading" } | { type: "value", value: T } | { type: "error", value: TError };

// will still throw on error
const suspendable = <T>(atom<Promise<T> | T>) => T | SUSPENDED;

// will not throw on error
const loadable = <T>(atom<Promise<T> | T>) => Loadable<T>;
```

Personally, I'd be using the `suspendable` variant.

**Comment by dai-shi** [comment]
@Pinpickle 

> I'd love to have a go working on it! I think I might be able to sometime in the next couple of weeks.

Sounds great! I have a basic idea for the impl, so you can ask help if needed.
(tests and docs take time, so it's very helpful, if you can work on it. but feel free to let me know when you get busy.)

Would you please create a new issue to track and discuss about this?



> We could make `loadable` cache the created atom in a `WeakMap`, so this becomes possible:

Exactly. We do this as much as possible, see `selectAtom` for example.

> What do you think about also having a `useLoadableAtomValue` hook that automatically creates a

We don't do that. We focus on utils on atom level, not hook level. (We prefer minimal/primitive API.)

```js
const [slice, setSlice] = useAtom(selectAtom(anAtom, selector));

// if developers want they can create a custom hook on their end
const useSelector = (anAtom, selector) => useAtom(selectAtom(anAtom, selector))
```

> Personally I'm not sure about the error boundary use-case, I can image it being useful to throw or not throw depending on context. What about having two different functions?

Hmm, I see. I assumed they are mostly paired. Okay, sounds fair. I prefer single function with an option, though.

```js
loadable(anAtom, throwsError?: boolean) // you would use `true` for `throwsError`
```

We could probably keep the stale value, and returns stable object.

The typing would be a bit complicated. Roughly:

```tsx
function loadable<Value, Update>(
  atom: WritableAtom<Value, Update>
): WritableAtom<{ loading: true; data?: Value } | { data: Value } | { error: unknown }, Update>

function loadable<Value, Update>(
  atom: WritableAtom<Value, Update>,
  throwsError: boolean
): WritableAtom<{ loading: true; data?: Value } | { data: Value }, Update>

function loadable<Value>(
  atom: Atom<Value>
): Atom<{ loading: true; data?: Value } | { data: Value } | { error: unknown }>

function loadable<Value, Update>(
  atom: Atom<Value, Update>,
  throwsError: boolean
): Atom<{ loading: true; data?: Value } | { data: Value }>
```

`{ loading: boolean; data?: Value }` might be easier to use. returning loading: false.

**Comment by Pinpickle** [comment]
> > What do you think about also having a `useLoadableAtomValue` hook that automatically creates a
> 
> We don't do that. We focus on utils on atom level, not hook level. (We prefer minimal/primitive API.)

Oops, this sentence wasn't supposed to make it in, I thought I edited it out. I completely agree.

The nice thing about the `suspendable` API is that it's just a simple type union with the `SUSPENDED` symbol, which makes it just a little bit simpler to deal with.

I'll make a new issue so we can discuss this

**Comment by Pinpickle** [comment]
New issue: #672
